packages/theme/src/cookie-theme-provider.tsx

```diff
@@ -0,0 +1,113 @@
+import React, {
+  createContext,
+  FC,
+  PropsWithChildren,
+  useContext,
+  useEffect,
+  useMemo,
+  useState,
+} from 'react'
+import {
+  DEFAULT_THEME_NAME,
+  prefersDarkThemeMediaQuery,
+  ThemeName,
+} from './constants'
+import { themeMap } from './themes'
+import { getThemeNameFromCookies } from './utils/cookies'
+import { initColors } from './document-head-contents'
+import { updateGlobalTheme } from './document-head-contents/element-theme-script'
+import { ThemeProvider as StyledThemeProvider } from 'styled-components'
+import { ThemeContext } from './types'
+
+const defaultThemeContext = {} as ThemeContext
+export const ThemeToggleContext =
+  createContext<ThemeContext>(defaultThemeContext)
+
+initColors()
+
+/** Cookie theme provider acts differently from common theme provider.
+ * 1. it can be nested, BUT it should re-use value provided by itself for optimisation purposes
+ * 2. if it is top-level, it is not altering itself, but instead altering theme behavior on HTMLElement level
+ * 3. it does not support custom themes, preferring pre-defined themes instead
+ * */
+
+export const CookieThemeProvider: FC<
+  PropsWithChildren<{
+    // Use themeNameParent if you need get cookie in SSR
+    initialTheme?: ThemeName
+    overrideTheme?: ThemeName
+  }>
+> = React.memo(
+  ({
+    children,
+    initialTheme,
+    // overrideTheme is mainly used for storybook
+    overrideTheme,
+  }) => {
+    const parentTheme = useContext(ThemeToggleContext)
```

Comment:
Why do we import theme from ThemeToggle? From the name, I can only assume it's only toggle theme, but not export it

Reply:
this is what we're providing in this component, so if something is provided, it's nested